Mathlib/NumberTheory/FLT/MasonStothers.lean

```diff
@@ -0,0 +1,253 @@
+-/
+theorem divRadical_dvd_derivative (a : k[X]) : (divRadical a) ∣ (derivative a) := by
+  induction a using induction_on_coprime
+  · case h0 =>
+    rw [derivative_zero]
+    apply dvd_zero
+  · case h1 a ha =>
+    exact (divRadical_isUnit ha).dvd
+  · case hpr p i hp =>
+    cases i
+    · rw [pow_zero, derivative_one]
+      apply dvd_zero
+    · case succ i =>
+      rw [← mul_dvd_mul_iff_left (radical_ne_zero (p ^ i.succ)), hMul_radical_divRadical,
+        radical_pow_of_prime hp i.succ_pos, derivative_pow_succ, ← mul_assoc]
+      apply dvd_mul_of_dvd_left
+      rw [mul_comm, mul_assoc]
+      apply dvd_mul_of_dvd_right
+      rw [pow_succ, mul_dvd_mul_iff_left (pow_ne_zero i hp.ne_zero), dvd_normalize_iff]
+  · -- If it holds for coprime pair a and b, then it also holds for a * b.
+    case hcp x y hpxy hx hy =>
+    have hc : IsCoprime x y :=
+      EuclideanDomain.isCoprime_of_dvd
+        (fun ⟨hx, hy⟩ => not_isUnit_zero (hpxy (zero_dvd_iff.mpr hx) (zero_dvd_iff.mpr hy)))
+        fun p hp _ hpx hpy => hp (hpxy hpx hpy)
+    rw [divRadical_hMul hc, derivative_mul]
+    exact dvd_add (mul_dvd_mul hx (divRadical_dvd_self y)) (mul_dvd_mul (divRadical_dvd_self x) hy)
+
+theorem divRadical_dvd_wronskian_left (a b : k[X]) : (divRadical a) ∣ wronskian a b := by
+  rw [wronskian]
+  apply dvd_sub
+  · apply dvd_mul_of_dvd_left
+    exact divRadical_dvd_self a
+  · apply dvd_mul_of_dvd_left
+    exact divRadical_dvd_derivative a
+
+theorem divRadical_dvd_wronskian_right (a b : k[X]) : (divRadical b) ∣ wronskian a b := by
+  rw [← wronskian_neg_eq, dvd_neg]
+  exact divRadical_dvd_wronskian_left _ _
+
+@[simp]
+theorem dvd_derivative_iff {a : k[X]} : a ∣ derivative a ↔ derivative a = 0 := by
+  constructor
+  · intro h
+    by_cases a_nz : a = 0
+    · rw [a_nz]; simp only [derivative_zero]
+    by_contra deriv_nz
+    have deriv_lt := degree_derivative_lt a_nz
+    have le_deriv := Polynomial.degree_le_of_dvd h deriv_nz
+    have lt_self := le_deriv.trans_lt deriv_lt
+    simp only [lt_self_iff_false] at lt_self
+  intro h; rw [h]; simp
+
+theorem IsCoprime.wronskian_eq_zero_iff {a b : k[X]} (hc : IsCoprime a b) :
+    wronskian a b = 0 ↔ derivative a = 0 ∧ derivative b = 0 := by
+  constructor
+  · intro hw
+    rw [wronskian, sub_eq_iff_eq_add, zero_add] at hw
+    constructor
+    · rw [← dvd_derivative_iff]
+      apply hc.dvd_of_dvd_mul_right
+      rw [← hw]; exact dvd_mul_right _ _
+    · rw [← dvd_derivative_iff]
+      apply hc.symm.dvd_of_dvd_mul_left
+      rw [hw]; exact dvd_mul_left _ _
+  intro hdab
+  cases' hdab with hda hdb
+  rw [wronskian]
+  rw [hda, hdb]; simp only [MulZeroClass.mul_zero, MulZeroClass.zero_mul, sub_self]
+
+
+protected theorem IsCoprime.divRadical {a b : k[X]} (h : IsCoprime a b) :
+    IsCoprime (divRadical a) (divRadical b) := by
+  rw [← hMul_radical_divRadical a] at h
+  rw [← hMul_radical_divRadical b] at h
+  exact h.of_mul_left_right.of_mul_right_right
```

Comment:
Could all of this move to a file under `Mathlib.RingTheory.Polynomial`? Either `Mathlib.RingTheory.Polynomial.Wronskian` or a new file `Mathlib.RingTheory.Polynomial.DivRadical`? And can you make that into a separate PR?

Reply:
I'll put under `Mathlib.RingTheory.Polynomial` then (IMO these theorems are to specific so I'm not sure if it worths to make new files for it).

Reply:
I just realized that first half of the theorems & definition of `divRadical` actually makes sense for `EuclideanDomain` as well, but not sure if I put these under the "euclidean domain" files. Do you have any suggestions? I think another way is to put `divRadical`-related theorems, up to `divRadical_dvd_self`, in `Mathlib.RingTheory.Radical` with `EuclideanDomain` version, and put the rest in `Mathlib.RingTheory.Polynomial`? (Honestly I've never seen any mathematical proofs using $x / \mathrm{rad}(x)$ except for Mason-Stothers...)

Reply:
For `dvd_derivative_iff` and `IsCoprime.wronskian_eq_zero_iff`, I think these are also seperate things. I'd put the second one under `Mathlib.RingTheory.Polynomial.Wronskian`, but this depends on the first theorem and I don't know where's the best place to put it - do you have any suggestions? (these two are not included in #18452 for now)

Reply:
No sorry, no suggesyion. I would play around with the myself but I am momentarily incapactitated by apainful wrist